File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        // we need to check for rows; which row contains the value 
        int rows = matrix.length;
        int cols = matrix[0].length;
        int targetRow = 0;
        for(targetRow = 0; targetRow< rows; targetRow++){
            if(target >= matrix[targetRow][0] && target <= matrix[targetRow][cols -1]){
                break;
            }
        }
        if(targetRow >= 0 && targetRow < rows){
            //start the bianry search just on this row
            int left = 0, right = cols - 1;

            while(left <= right){
               int mid = left + (right - left)/2;

                if(target == matrix[targetRow][mid]){
                    return true;
                }else if(target > matrix[targetRow][mid]){
                    left = mid + 1;
                }else{
                    right = mid - 1;
                }
            }
           
        }
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
       let mut rows = matrix.len();
       let mut cols = matrix[0].len();

       //find the target_row
       let mut target_row = 0;
       for row in 0..rows{
        if(matrix[row][0] <= target && matrix[row][cols - 1] >= target){
            target_row = row;
        }
       } 
       println!("{}", target_row);
       if target_row >= rows || target_row < 0{
        return false;
       }
       println!("{}", target_row);
       //find the lelemnt in target_row
       let mut left = 0;
       let mut right = cols;
       while(left < right){
            let mid = left + (right - left)/2;
            if matrix[target_row][mid] == target{
                return true;
            }
            if matrix[target_row][mid] >= target{
                right = mid;
            }else{
                left = mid + 1;
            }
       }
       false
    }
}